presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -492,6 +493,99 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
+        {
+            Expression inheritedPredicate = context.get();
+
+            // See if we can rewrite left join in terms of a plain inner join
+            if (node.getType() == SpatialJoinNode.Type.LEFT && canConvertOuterToInner(node.getRight().getOutputSymbols(), inheritedPredicate)) {
+                node = new SpatialJoinNode(node.getId(), SpatialJoinNode.Type.INNER, node.getLeft(), node.getRight(), node.getOutputSymbols(), node.getFilter());
+            }
+
+            Expression leftEffectivePredicate = effectivePredicateExtractor.extract(node.getLeft());
+            Expression rightEffectivePredicate = effectivePredicateExtractor.extract(node.getRight());
+            Expression joinPredicate = node.getFilter();
+
+            Expression leftPredicate;
+            Expression rightPredicate;
+            Expression postJoinPredicate;
+            Expression newJoinPredicate;
+
+            switch (node.getType()) {
+                case INNER:
+                    InnerJoinPushDownResult innerJoinPushDownResult = processInnerJoin(inheritedPredicate,
+                            leftEffectivePredicate,
+                            rightEffectivePredicate,
+                            joinPredicate,
+                            node.getLeft().getOutputSymbols());
+                    leftPredicate = innerJoinPushDownResult.getLeftPredicate();
+                    rightPredicate = innerJoinPushDownResult.getRightPredicate();
+                    postJoinPredicate = innerJoinPushDownResult.getPostJoinPredicate();
+                    newJoinPredicate = innerJoinPushDownResult.getJoinPredicate();
+                    break;
+                case LEFT:
+                    OuterJoinPushDownResult leftOuterJoinPushDownResult = processLimitedOuterJoin(inheritedPredicate,
+                            leftEffectivePredicate,
+                            rightEffectivePredicate,
+                            joinPredicate,
+                            node.getLeft().getOutputSymbols());
+                    leftPredicate = leftOuterJoinPushDownResult.getOuterJoinPredicate();
+                    rightPredicate = leftOuterJoinPushDownResult.getInnerJoinPredicate();
+                    postJoinPredicate = leftOuterJoinPushDownResult.getPostJoinPredicate();
+                    newJoinPredicate = leftOuterJoinPushDownResult.getJoinPredicate();
+                    break;
+                default:
+                    throw new UnsupportedOperationException("Unsupported spatial join type: " + node.getType());
+            }
+
+            newJoinPredicate = simplifyExpression(newJoinPredicate);
+            // TODO: find a better way to directly optimize FALSE LITERAL in join predicate
+            if (newJoinPredicate.equals(BooleanLiteral.FALSE_LITERAL)) {
+                newJoinPredicate = new ComparisonExpression(ComparisonExpression.Operator.EQUAL, new LongLiteral("0"), new LongLiteral("1"));
```

Comment:
why can't we leave `FALSE_LITERAL` here?

Reply:
@kokosing I don't really know. This code was added in https://github.com/prestodb/presto/pull/4716/files to fix #4712 . @martint Martin, do you know why this check is needed? The fix added `AbstractTestQueries#testJoinWithInferredFalseJoinClause` test, so I tried commenting out this code and running the test. The test succeeded. Perhaps, this check is no longer needed?

Reply:
I believe this is from the time when we didn't support non-equi joins.